File: javascript/momoization/countConstruct.js
const countConstruct = (target, wordBank, memo = {}) => {
    if (target in memo) return memo[target];
    if (target === '') return 1;

    let totalCount = 0;

    for (let word of wordBank) {
        if (target.indexOf(word) === 0) {
           const numWaysForRest = countConstruct(target.slice(word.length), wordBank, memo);
            totalCount += numWaysForRest;
        }
    }

    memo[target] = totalCount;
    return totalCount
}

console.log(countConstruct('purple', ['p', 'urple', 'purp', 'le']))

File: ruby/tabulation/countConstruct.rb
def count_construct(target, wordBank)
    table = Array.new(target.length + 1, 0)

    table[0] = 1

    (target.length + 1).times do |i|
        wordBank.each do |w|
            if target.slice(i, w.length) === w
                table[i+w.length] += table[i]
            end
        end
    end

    table[target.length]
end

p count_construct('purple', ['purp', 'le', 'pur', 'ple'])